core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -0,0 +1,387 @@
+  import ProducerIdMapping._
+
+  val snapDir: File = new File(snapParentDir, formatDirName(topicPartition))
+  snapDir.mkdir()
+
+  private val pidMap = mutable.Map[Long, ProducerIdEntry]()
+  private var lastMapOffset = -1L
+  private var lastSnapOffset = -1L
+
+  /**
+   * Returns the last offset of this map
+   */
+  def mapEndOffset = lastMapOffset
+
+  /**
+   * Get a copy of the active producers
+   */
+  def activePids: immutable.Map[Long, ProducerIdEntry] = pidMap.toMap
+
+  /**
+   * Load a snapshot of the id mapping or return empty maps
+   * in the case the snapshot doesn't exist (first time).
+   */
+  private def loadFromSnapshot(logEndOffset: Long) {
+    pidMap.clear()
+    lastSnapshotFile(logEndOffset) match {
+      case Some(file) =>
+        try {
+          trace("Loading snapshot from file: %s".format(file.toPath))
+          loadSnapshot(file, pidMap)
+          lastSnapOffset = offsetFromFile(file)
+          lastMapOffset = lastSnapOffset
+        } catch {
+          case e: CorruptSnapshotException =>
+            // TODO: Delete the file and try to use the next snapshot
+            throw e
+        }
+
+      case None =>
+        lastSnapOffset = -1
+        lastMapOffset = -1
+        snapDir.mkdir()
+    }
+  }
+
+  def checkForExpiredPids(currentTimeMs: Long) {
+    pidMap.retain { case (pid, lastEntry) =>
+      currentTimeMs - lastEntry.timestamp < maxPidExpirationMs
+    }
+  }
+
+  def truncateAndReload(logEndOffset: Long) = {
+    truncateSnapshotFiles(logEndOffset)
+    loadFromSnapshot(logEndOffset)
+  }
+
+  /**
+   * Update the mapping with the given append information
+   */
+  def update(appendInfo: ProducerAppendInfo): Unit = {
+    if (appendInfo.pid == RecordBatch.NO_PRODUCER_ID)
+      throw new IllegalArgumentException("Invalid PID passed to update")
+    val entry = appendInfo.lastEntry
+    pidMap.put(appendInfo.pid, entry)
+    lastMapOffset = entry.lastOffset
+  }
+
+  /**
+   * Load a previously stored PID entry into the cache. Ignore the entry if the timestamp is older
+   * than the current time minus the PID expiration time (i.e. if the PID has expired).
+   */
+  def load(pid: Long, entry: ProducerIdEntry, currentTimeMs: Long) {
+    if (pid != RecordBatch.NO_PRODUCER_ID && currentTimeMs - entry.timestamp < maxPidExpirationMs) {
+      pidMap.put(pid, entry)
+      lastMapOffset = entry.lastOffset
+    }
+  }
+
+  /**
+   * Get the last written entry for the given PID.
+   */
+  def lastEntry(pid: Long): Option[ProducerIdEntry] = pidMap.get(pid)
+
+  /**
+    * Serialize and write the bytes to a file. The file name is a concatenation of:
+    *   - offset
+    *   - a ".snapshot" suffix
+    *
+    * The parent directory is a concatenation of:
+    *   - log.dir
+    *   - base name
+    *   - topic
+    *   - partition id
+    *
+    *   The offset is supposed to increase over time and
+    */
+  def maybeTakeSnapshot() {
```

Comment:
It doesn't seem that we have the logic to take snapshots periodically since this method is only called from tests?